src/image.rs

```diff
@@ -147,3 +154,180 @@ impl Image {
         self
     }
 }
+
+/// Describes the image content of a filled or stroked shape.
+///
+/// See also [`ImageBrushRef`] which can be used to avoid allocations.
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct ImageBrush {
+    /// The image to render.
+    pub image: ImageData,
+    /// Parameters which specify how to render the image.
+    pub params: ImageRenderParams,
+}
+
+impl ImageBrush {
+    /// Creates a new image with the given data, [format](ImageFormat) and dimensions.
+    #[must_use]
+    pub fn new(image: ImageData) -> Self {
+        Self {
+            image,
+            params: ImageRenderParams::default(),
+        }
+    }
+
+    /// Converts the `ImageBrushRef` to an owned `ImageBrush`.
+    #[must_use]
+    pub fn as_ref(&self) -> ImageBrushRef<'_> {
+        ImageBrushRef {
+            image: &self.image,
+            params: self.params,
+        }
+    }
+
+    /// Builder method for setting the image [extend mode](Extend) in both
+    /// directions.
+    #[must_use]
+    pub fn with_extend(mut self, mode: Extend) -> Self {
+        self.params.x_extend = mode;
+        self.params.y_extend = mode;
+        self
+    }
+
+    /// Builder method for setting the image [extend mode](Extend) in the
+    /// horizontal direction.
+    #[must_use]
+    pub fn with_x_extend(mut self, mode: Extend) -> Self {
+        self.params.x_extend = mode;
+        self
+    }
+
+    /// Builder method for setting the image [extend mode](Extend) in the
+    /// vertical direction.
+    #[must_use]
+    pub fn with_y_extend(mut self, mode: Extend) -> Self {
+        self.params.y_extend = mode;
+        self
+    }
+
+    /// Builder method for setting a hint for the desired image [quality](ImageQuality)
+    /// when rendering.
+    #[must_use]
+    pub fn with_quality(mut self, quality: ImageQuality) -> Self {
+        self.params.quality = quality;
+        self
+    }
+
+    /// Returns the image with the alpha multiplier set to `alpha`.
+    #[must_use]
+    #[track_caller]
+    pub fn with_alpha(mut self, alpha: f32) -> Self {
+        debug_assert!(
+            alpha.is_finite() && alpha >= 0.0,
+            "A non-finite or negative alpha ({alpha}) is meaningless."
+        );
+        self.params.alpha = alpha;
+        self
+    }
+
+    /// Returns the image with the alpha multiplier multiplied again by `alpha`.
+    /// The behaviour of this transformation is undefined if `alpha` is negative.
+    #[must_use]
+    #[track_caller]
+    pub fn multiply_alpha(mut self, alpha: f32) -> Self {
+        debug_assert!(
+            alpha.is_finite() && alpha >= 0.0,
+            "A non-finite or negative alpha ({alpha}) is meaningless."
+        );
+        self.params.alpha *= alpha;
+        self
+    }
+}
+
+/// Borrowed version of [`ImageBrush`] for avoiding allocations
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct ImageBrushRef<'a> {
```

Comment:
It's not clear what allocations this is avoiding. `Image` is refcounted, so cloning it doesn't cause an allocation.

Reply:
Atomic refcount bumps can still be expensive if there are a lot of them

Reply:
Sure, that is entirely true. However, I still want the docs changed to not claim that it's "for avoiding allocations" when it isn't.

Separately, I do think that it's likely that anything which uses this type will just clone the image straight away, and so this type is just adding more cognitive load for no gain, and so giving false impressions. I'd be happier with this PR if it were removed, but I'm not going to block on that.